scipy/interpolate/_bsplines.py

```diff
@@ -0,0 +1,879 @@
+
+    >>> x = cheb_nodes(20)
+    >>> y = np.sqrt(1 - x**2)
+
+    >>> from scipy.interpolate import BSpline, make_interp_spline
+    >>> b = make_interp_spline(x, y)
+    >>> np.allclose(b(x), y)
+    True
+
+    Note that the default is a cubic spline with a not-a-knot boundary condition
+
+    >>> b.k
+    3
+
+    Here we use a 'natural' spline, with zero 2nd derivatives at edges:
+
+    >>> l, r = [(2, 0)], [(2, 0)]
+    >>> b_n = make_interp_spline(x, y, deriv_l=l, deriv_r=r)
+    >>> np.allclose(b_n(x), y)
+    True
+    >>> x0, x1 = x[0], x[-1]
+    >>> np.allclose([b_n(x0, 2), b_n(x1, 2)], [0, 0])
+    True
+
+    Interpolation of parametric curves is also supported. As an example, we
+    compute a discretization of a snail curve in polar coordinates
+
+    >>> phi = np.linspace(0, 2.*np.pi, 40)
+    >>> r = 0.3 + np.cos(phi)
+    >>> x, y = r*np.cos(phi), r*np.sin(phi)  # convert to Cartesian coordinates
+
+    Build an interpolating curve, parameterizing it by the angle
+
+    >>> from scipy.interpolate import make_interp_spline
+    >>> spl = make_interp_spline(phi, np.c_[x, y])
+
+    Evaluate the interpolant on a finer grid (note that we transpose the result
+    to unpack it into a pair of x- and y-arrays)
+
+    >>> phi_new = np.linspace(0, 2.*np.pi, 100)
+    >>> x_new, y_new = spl(phi_new).T
+
+    Plot the result
+
+    >>> import matplotlib.pyplot as plt
+    >>> plt.plot(x, y, 'o')
+    >>> plt.plot(x_new, y_new, '-')
+    >>> plt.show()
+
+    See Also
+    --------
+    BSpline : base class representing the B-spline objects
+    make_lsq_spline : a similar factory function for spline fitting
+    UnivariateSpline : a wrapper over FITPACK spline fitting routines
+    splrep : a wrapper over FITPACK spline fitting routines
+
+    """
+    # special-case k=0 right away
+    if k == 0:
+        if any(_ is not None for _ in (t, deriv_l, deriv_r)):
+            raise ValueError("Too much info for k=0.")
+        t = np.r_[x, x[-1]]
+        c = y
+        return BSpline(t, c, k, axis=axis)
+
+    # come up with a sensible knot vector, if needed
+    if t is None:
+        if deriv_l is None and deriv_r is None:
+            if k == 2:
+                # OK, it's a bit ad hoc: Greville sites + omit
+                # 2nd and 2nd-to-last points, a la not-a-knot
+                t = (x[1:] + x[:-1]) / 2.
+                t = np.r_[(x[0],)*(k+1),
+                           t[1:-1],
+                           (x[-1],)*(k+1)]
+            else:
+                t = _not_a_knot(x, k)
+        else:
+            t = _augknt(x, k)
+
+    x = _as_float_array(x, check_finite)
+    y = _as_float_array(y, check_finite)
+    t = _as_float_array(t, check_finite)
+    k = int(k)
+
+    axis = axis % y.ndim
+    y = np.rollaxis(y, axis)    # now internally interp axis is zero
+
+    if x.ndim != 1 or np.any(x[1:] - x[:-1] <= 0):
+        raise ValueError("Expect x to be a 1-D sorted array_like.")
+    if x.shape[0] < k+1:
```

Comment:
Hello,

This is the first time I make a pull request comment, so I accept any guidance on how to make this more effectively.

This test:
```python
if x.shape[0] < k+1:
    raise ValueError("Need more x points.")
```
is not necessary (I think) and prevent to use the function `make_interp_spline` in legitimate cases. For example if you interpolate 2 points with a cubic spline and that you impose left and right first derivatives, your spline is fully defined. (2 points + 2 derivatives = 4 conditions for the 4 control points required by a cubic spline). This use case is commonplace when you need to interpolate splines through points for geometric constructions (ala CAD software).

Note also that AFAICT if you do not define enough conditions (i.e. number of points and derivatives), the test L.686 will catch it:
```python
if nt - n != nleft + nright:
    raise ValueError("number of derivatives at boundaries.")
```
Finally, you can prove my point with the following test (with the incriminated test commented out):
```python
import numpy as np
from numpy.testing import assert_allclose
from scipy.interpolate import make_inter_spline

# interpolation of f(x) = x**3 => f'(0) = 0, f'(1) = 3
x = [0., 1.]
y = [0., 1.]
deriv_l = [(1, [0.])]
deriv_r = [(1, [3.])]
s = make_interp_spline(x, y, k=3, deriv_l=deriv_l, deriv_r=deriv_r)

xx = np.linspace(0., 1,.)
assert_allclose(xx**3, s(xx), rtol=1e-14)
```

Thanks